other/upgrade.php

```diff
@@ -1061,6 +1064,111 @@ function WelcomeLogin()
 	return false;
 }
 
+// Do a number of attachment & avatar folder checks.
+// Display a warning if issues found.  Does not force a hard stop.
+function checkFolders()
+{
+	global $modSettings, $upcontext, $txt, $command_line;
+
+	$warnings = '';
+
+	// First, check the avatar directory...
+	// Note it wasn't specified in yabbse, but there was no smfVersion either.
+	if (!empty($modSettings['smfVersion']) && !is_dir($modSettings['avatar_directory']))
+		$warnings .= $txt['warning_av_missing'];
+
+	// Next, check the custom avatar directory...  Note this is optional in 2.0.
+	if (!empty($modSettings['custom_avatar_dir']) && !is_dir($modSettings['custom_avatar_dir']))
+	{
+		if (empty($warnings))
+			$warnings = $txt['warning_custom_av_missing'];
+		else
+			$warnings .= '<br><br>' . $txt['warning_custom_av_missing'];
+	}
+
+	// Finally, attachment folders.
+	// A bit more complex, since it may be json or serialized, and it may be an array or just a string...
+	$ser_test = @unserialize($modSettings['attachmentUploadDir']);
+	$json_test = @json_decode($modSettings['attachmentUploadDir'], true);
+	$string_test = !empty($modSettings['attachmentUploadDir']) && is_string($modSettings['attachmentUploadDir']) && is_dir($modSettings['attachmentUploadDir']);
+
+	// String?
+	$attdr_problem_found = false;
+	if ($string_test === true)
+	{
+		// OK...
+	}
+	// Serialized?
+	elseif ($ser_test !== false)
+	{
+		if (is_array($ser_test))
+		{
+			foreach($ser_test AS $dir)
+			{
+				if (!is_dir($dir))
+					$attdr_problem_found = true;
+			}	
+		}
+		else
+		{
+			if (!is_dir($ser_test))
+				$attdr_problem_found = true;
+		}
+	}
+	// Json?  Note the test returns null if encoding was unsuccessful
+	elseif ($json_test !== null)
+	{
+		if (is_array($json_test))
+		{
+			foreach($json_test AS $dir)
+			{
+				if (!is_dir($dir))
+					$attdr_problem_found = true;
+			}	
+		}
+		else
+		{
+			if (!is_dir($json_test))
+				$attdr_problem_found = true;
+		}
+	}
```

Comment:
They all three seem to not depend on each other? It works though

Reply:
Not sure the question here.  The string is either plain text, serialized, or json.  Can't be more than one.

The only way I can reproduce the original issue (attachments/avatars not converted) is by starting with an incorrect attachment folder.